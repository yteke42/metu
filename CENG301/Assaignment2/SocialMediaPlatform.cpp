#include "SocialMediaPlatform.h"
#include "User.h"
#include "SortedLinkedList.h"
#include "LinkedListNode.h"

using namespace std;

// Implement functions in the "SocialMediaPlatform.h"
SocialMediaPlatform::SocialMediaPlatform() : users(new SortedLinkedList()) {}

SocialMediaPlatform::~SocialMediaPlatform() { delete users; }

bool SocialMediaPlatform::addUser(const int userId, const std::string name) {
    if (users->checkIfExist(userId)) return false;

    User *newUser = new User();

    newUser->setId(userId);
    newUser->setName(name);
    users->add(newUser);
    return true;
} 

bool SocialMediaPlatform::removeUser(const int userId) {
    // Iterate through all users in the platform
    LinkedListNode *currentUserNode = users->getHead();
    while (currentUserNode != nullptr) {
        // Remove the user to be deleted from the current user's followeds and followers list
        currentUserNode->u->removeFollowed(userId);
        currentUserNode->u->removeFollower(userId);

        // Move to the next user
        currentUserNode = currentUserNode->next;
    }

    // Remove the user from the platform
    return users->remove(userId);
}





bool SocialMediaPlatform::followUser(const int firstUserId, const int secondUserId) {
    LinkedListNode *firstUserNode = users->getPointerTo(firstUserId);
    LinkedListNode *secondUserNode = users->getPointerTo(secondUserId);

    if (firstUserNode == nullptr || secondUserNode == nullptr) return false;

    if (firstUserNode->u->getFolloweds()->checkIfExist(secondUserId)) {
        // User is already following the second user
        return false;
    }

    firstUserNode->u->addFollowed(secondUserNode->u);
    secondUserNode->u->addFollower(firstUserNode->u);
    return true;
}


bool SocialMediaPlatform::unfollowUser(const int firstUserId, const int secondUserId) {
    LinkedListNode *firstUserNode = users->getPointerTo(firstUserId);
    LinkedListNode *secondUserNode = users->getPointerTo(secondUserId);

    if (firstUserNode == nullptr || secondUserNode == nullptr) return false;

    if (!firstUserNode->u->getFolloweds()->remove(secondUserId) ||
        !secondUserNode->u->getFollowers()->remove(firstUserId)) {
        return false; // Either user not found or user not followed, return false
    }

    return true;
}


void SocialMediaPlatform::showFollowersOf(const int userId) const {
    LinkedListNode *userNode = users->getPointerTo(userId);
    if (userNode != nullptr) {
        std::cout << "Users following the user with ID " << userId << ":" << std::endl;
        if (!userNode->u->getFollowers()->isEmpty()) {
            userNode->u->getFollowers()->print();
        } else {
            std::cout << "None" << std::endl;
        }
    } else {
        std::cout << "Cannot show. There is no user with ID " << userId << "." << std::endl;
    }
}





void SocialMediaPlatform::showAllUsers() const {
    std::cout << "Users in the social media platform:" << std::endl;
    if (users->isEmpty()) {
        std::cout << "None" << std::endl;
    } else {
        users->print();
    }
}





