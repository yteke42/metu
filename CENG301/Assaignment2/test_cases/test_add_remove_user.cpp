#include<iostream>
#include<string>
#include "SocialMediaPlatform.h"

using namespace std;

int main() {

    string truth_table[2] = {"false", "true"};

    SocialMediaPlatform SMP;

    cout<<truth_table[SMP.addUser( -4, "Mert" )]<<endl;
    cout<<truth_table[SMP.addUser( 5, "Ferhat" )]<<endl;
    cout<<truth_table[SMP.addUser( 8, "Selin" )]<<endl;
    cout<<truth_table[SMP.addUser( 3, "Hatice" )]<<endl;

    SMP.showAllUsers();
    cout<<endl;
    
    cout<<truth_table[SMP.removeUser(3)]<<endl;
 
    SMP.showAllUsers();
    cout<<endl;
    
    cout<<truth_table[SMP.removeUser(-4)]<<endl;
    
    SMP.showAllUsers();
    cout<<endl;
    
    cout<<truth_table[SMP.removeUser(8)]<<endl;
    
    SMP.showAllUsers();
    cout<<endl;
    
    cout<<truth_table[SMP.removeUser(5)]<<endl;

    SMP.showAllUsers();
    cout<<endl;

    return 0;
}
