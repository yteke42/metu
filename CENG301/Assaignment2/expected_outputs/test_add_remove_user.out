true
true
true
true
Users in the social media platform:
-4, Mert
3, Hatice
5, Ferhat
8, Selin

true
Users in the social media platform:
-4, Mert
5, Ferhat
8, Selin

true
Users in the social media platform:
5, Ferhat
8, Selin

true
Users in the social media platform:
5, Ferhat

true
Users in the social media platform:
None

